src/BenchmarkDotNet.TestAdapter/VSTestAdapter.cs

```diff
@@ -0,0 +1,208 @@
+                var idLookup = new Dictionary<Guid, string>();
+                foreach (var testCase in testCases)
+                {
+                    if (idLookup.TryGetValue(testCase.Id, out var matchingCase))
+                        throw new Exception($"Encountered Duplicate Test ID: '{testCase.DisplayName}' and '{matchingCase}'");
+
+                    idLookup[testCase.Id] = testCase.DisplayName;
+                }
+
+                return testCases;
+            }
+            catch (Exception ex)
+            {
+                logger.SendMessage(TestMessageLevel.Error, $"Failed to load benchmarks from source\n{ex}");
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Runs the benchmarks in the given source.
+        /// </summary>
+        /// <param name="source">The dll or exe of the benchmark project.</param>
+        /// <param name="frameworkHandle">An interface used to communicate with the VSTest host.</param>
+        /// <param name="testCases">
+        /// The specific test cases to be run if specified.
+        /// If unspecified, runs all the test cases in the source.
+        /// </param>
+        private void RunBenchmarks(string source, IFrameworkHandle frameworkHandle, IEnumerable<TestCase>? testCases = null)
+        {
+            ValidateSourceOrThrow(source);
+
+            // Create a HashSet of all the TestCase IDs to be run if specified.
+            var caseIds = testCases == null ? null : new HashSet<Guid>(testCases.Select(c => c.Id));
+
+            try
+            {
+                // Ensure that test execution is done inside the context of the source directory.
+                var executor = (BenchmarkExecutorWrapper)CreateIsolatedType(typeof(BenchmarkExecutorWrapper), source);
+                cts?.Token.Register(executor.Cancel);
+
+                executor.RunBenchmarks(source, new TestExecutionRecorderWrapper(frameworkHandle), caseIds);
+            }
+            catch (Exception ex)
+            {
+                frameworkHandle.SendMessage(TestMessageLevel.Error, $"Failed to run benchmarks in source\n{ex}");
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// This will create the given type in a child AppDomain when used in .NET Framework.
+        /// If not in the .NET Framework, it will use the current
+        /// </summary>
+        /// <param name="type">The type to create.</param>
+        /// <param name="source">The dll or exe of the benchmark project.</param>
+        /// <returns>The created object.</returns>
+        private static object CreateIsolatedType(Type type, string source)
+        {
+#if NETFRAMEWORK
+            var appBase = Path.GetDirectoryName(source);
+            var setup = new AppDomainSetup { ApplicationBase = appBase };
+            var domainName = $"Isolated Domain for {type.Name}";
+            var appDomain = AppDomain.CreateDomain(domainName, null, setup);
```

Comment:
Is it really necessary? What do we benefit out of creating a dedicated app domain?

Reply:
Without this, it is impossible to get .NET Framework to resolve the correct assemblies as otherwise it will try load them from the directory that the vstest console executable lives in rather than the BenchmarkDotNet folder. .NET Core is smarter around assembly resolution and does not need this. The same approach of using an AppDomain to load the assemblies for .NET Framework is also used by MSTest and XUnit libraries, so I fear that it may be the only option.